edward/inferences/replica_exchange_mc.py

```diff
@@ -0,0 +1,284 @@
+                  for beta in betas]
+
+    # Make replica.
+    self.replica_vars = []
+    for beta in self.betas:
+      self.replica_vars.append({z: Empirical(params=tf.Variable(tf.zeros(
+          qz.params.shape, dtype=latent_vars[z].dtype))) for z, qz in
+          six.iteritems(latent_vars)})
+
+    self.exchange_freq = exchange_freq
+
+    super(ReplicaExchangeMC, self).__init__(latent_vars, data)
+
+  def initialize(self, *args, **kwargs):
+    kwargs['auto_transform'] = False
+    return super(ReplicaExchangeMC, self).initialize(*args, **kwargs)
+
+  def build_update(self):
+    """
+    Perform sampling and exchange.
+    """
+    # Sample by Metropolis-Hastings for each replica.
+    replica_sample = []
+    replica_accept = []
+    for i, beta in enumerate(self.betas):
+      sample_, accept_ = self.mh_sample(self.replica_vars[i], beta)
+      replica_sample.append(sample_)
+      replica_accept.append(accept_)
+    accept = replica_accept[0]
+
+    # Variable to store order of replicas after exchange
+    new_replica_idx = tf.Variable(tf.range(self.n_replica))
+    new_replica_idx = tf.assign(new_replica_idx, tf.range(self.n_replica))
+
+    # Exchange adjacent replicas at frequency of exchange_freq
+    i = tf.random_uniform((), maxval=2, dtype=tf.int32)
+
+    def c(i, new_replica_idx):
+        return tf.less(i, self.n_replica - 1)
+
+    def b(i, new_replica_idx):
+        return [tf.add(i, 2), self.replica_exchange(i, i + 1, replica_sample,
+                                                    new_replica_idx)]
+
+    def exchange_all():
+        return tf.while_loop(c, b, loop_vars=[i, new_replica_idx])
+
+    u = tf.random_uniform([])
+    exchange = u < self.exchange_freq
+    i, new_replica_idx = tf.cond(exchange,
+                                 exchange_all,
+                                 lambda: [i, new_replica_idx])
+
+    # New replica sorted by new_replica_idx
+    new_replica_sample = []
+    for i in range(self.n_replica):
+      new_replica_sample.append(tf.case(
+          {tf.equal(tf.gather(new_replica_idx, i), j):
+           memory_lambda(replica_sample[j])
+           for j in range(self.n_replica)}, exclusive=True))
+
+    assign_ops = []
+
+    # Update Empirical random variables.
+    for z, qz in six.iteritems(self.latent_vars):
+      variable = qz.get_variables()[0]
+      assign_ops.append(tf.scatter_update(variable, self.t,
+                                          new_replica_sample[0][z]))
+
+    for i, beta in enumerate(self.betas):
+      for z, qz in six.iteritems(self.replica_vars[i]):
+        variable = qz.get_variables()[0]
+        assign_ops.append(tf.scatter_update(variable, self.t,
+                                            new_replica_sample[i][z]))
+
+    # Increment n_accept (if accepted).
+    assign_ops.append(self.n_accept.assign_add(tf.where(accept, 1, 0)))
+
+    return tf.group(*assign_ops)
+
+  def mh_sample(self, latent_vars, beta):
```

Comment:
rename to private method `_mh_sample`

Also, since this is a direct copy of MH algorithm's `build_update`, does it make sense to just build this class by inheriting from `MetropolisHastings`?

Reply:
Thanks!

As you say, except for using inverse temperature `beta`, this code is a copy of a part of MH algorithm's `build_update`. However in MH algorithm's `build_update`, sampling and updating are perform in the same function. In replica exchange method, exchanging must be perform between sampling and updating. Therefore, I think it is difficult to make `ReplicaExchangeMC` by inheriting from current `MetropolisHastings`. 

To make `ReplicaExchangeMC` by inheriting from current `MetropolisHastings`, 
It may be necessary to modify `MetropolisHastings`. (ex. creating `_mh_sample`)

And, replica exchange monte calro method doesn't necessarily have to use MH algorithm. Instead of MH algorithm, we can use Hamiltonian Monte Carlo and so on. So if possible, I'd like to make monte calro algorithms to use in `ReplicaExchangeMC` selectable.

Thank you for reading.
It would be extremely helpful if I could receive your advice.

Reply:
I see. Your choice is great—let's leave as is for now. We can always do refactoring in a future PR to avoid that code duplication.

Reply:
Thanks! I understand.